R/missingdata.R

```diff
@@ -173,4 +173,77 @@ ErrorIfInfinity <- function(data)
     }
 }
 
+#' @name AddDummyVariablesForNAs
+#'
+#' @title Add dummy variables to a \code{data.frame} suitable for regression models.
+#'
+#' @description Appends a matrix to a \code{data.frame} containing columns of dummy variables. It assumes
+#'   the input data is suitable for regression and contains a single column for an outcome variable.
+#'   Other columns would be for individual predictors used in the regression. The appended columns of dummy
+#'   variables indicate the missing status of each predictor. i.e. 1 if the predictor is missing in that case,
+#'   zero otherwise. A column is added for each predictor that has at least one case of missing data.
+#'
+#' @param data A \code{data.frame} of the data to be used for a Regression model assuming a single outcome.
+#' @param outcome.name A \code{characater} of the name of the outcome variable in the data.
+#' @param checks A \code{logical} to determine if further checks are done to remove cases from the data.
+#'   If \code{TRUE}, cases are removed if all predictors are missing or if the response is missing.
+#' @export
+AddDummyVariablesForNAs <- function(data, outcome.name, checks = TRUE)
+{
+    outcome <- data[which(names(data) == outcome.name)]
+    predictor.df <- data[-which(names(data) == outcome.name)]
+    # Create dummy variable matrix, only create column if necessary
+    dummy.variable.df <- lapply(predictor.df, function(x) {
+        z <- is.na(x)
+        if (any(z))
+            return(as.integer(z))
+        else
+            return(NULL)
+        })
+    # Remove the NULL elements (no missing)
+    dummy.variable.df <- Filter(Negate(is.null), dummy.variable.df)
+    missing.outcomes <- is.na(outcome[[1]])
+    # If no missing data in predictors, return original data, trimming missing outcomes if req
+    if (all(sapply(dummy.variable.df, is.null)))
```

Comment:
If NULLs were removed from `dummy.variable.df` on line 204, why do we check for them here?

Reply:
This is to catch the no missing predictors case. If there are no missing predictors, it returns the original data or data with missing outcome cases removed. I agree that if condition should be better. Perhaps `length(dummy.variable.df) == 0` instead?

Reply:
Could you change it to `length(dummy.variable.df) == 0`, just so that it is less confusing